### PR TITLE
Finished with challenges

### PR DESCRIPTION
**Pull request by rankdjr**

### CODE REVIEW

File: Flashcards.DouglasFir/Flashcards/App.config [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+<configuration>
+	<appSettings>
+		<!--' Database Constants | Database, Table, and View names '-->
+		<add key="DbName" value="FLASHCARDS" />
+		<add key="TblStackName" value="tb_Stacks" />
+		<add key="TblFlashCardsName" value="tb_FlashCards" />
+		<add key="TblStudySessionsName" value="tb_StudySessions" />
+		<add key="VwFlashCardsName" value="vw_FlashCardsDisplay" />
+		<add key="VwStudySessionsName" value="vw_StudySessionsWithStacks" />
+		<add key="VwFlashCardsRenumberedName" value="vw_FlashCardsRenumbered" />
```
Comment (antonderegt) [inline Flashcards.DouglasFir/Flashcards/App.config:11]: Please use readable names. I can guess tbl means table, but I don't know what vw stands for. Have a look at point 4 of the [Code Conventions](https://thecsharpacademy.com/article/58/code-conventions).

File: Flashcards.DouglasFir/Flashcards/App1.config [hunk 2]
```diff
@@ -0,0 +1,16 @@
+﻿<?xml version="1.0" encoding="utf-8" ?>
```
Comment (antonderegt) [inline Flashcards.DouglasFir/Flashcards/App1.config:1]: What's up with this file? It doesn't seem to be used and in the `.csproj` it gets removed.

File: Flashcards.DouglasFir/Flashcards/Application/AppStudySession.cs [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+        {
+            AnsiConsole.WriteLine("No data available for the selected year.");
+            _inputHandler.PauseForContinueInput();
+            return;
+        }
+
+        DisplayStudySessionReport(data);
+    }
+
+    private void DisplayStudySessionReport(IEnumerable<dynamic> data)
```
Comment (antonderegt) [inline Flashcards.DouglasFir/Flashcards/Application/AppStudySession.cs:153]: I would suggest to only use `dynamic` when really needed. Here you know what the data will look like.

File: Flashcards.DouglasFir/Flashcards/Application/AppStudySession.cs [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+        table.AddColumn("Aug");
+        table.AddColumn("Sep");
+        table.AddColumn("Oct");
+        table.AddColumn("Nov");
+        table.AddColumn("Dec");
+
+        foreach (var row in data)
+        {
+            string stackName = row.StackName;
+            string jan = row.Jan?.ToString() ?? "0";
```
Comment (antonderegt) [inline Flashcards.DouglasFir/Flashcards/Application/AppStudySession.cs:175]: Smart move to set the "0" value here. You don't have to change it, but just to make you aware, it's also possible to set the default value to 0 with SQL. Take a look at [COALESCE](https://www.w3schools.com/sql/func_sqlserver_coalesce.asp)

File: Flashcards.DouglasFir/Flashcards/DAO/FlashCardDao.cs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+            using (var dbConnection = _dbContext.GetConnectionToFlashCards())
+            {
+                string sql = $"INSERT INTO {ConfigSettings.tbFlashCardsName} (StackID, Front, Back) VALUES (@StackID, @Front, @Back)";
+                dbConnection.Execute(sql, new { flashCard.StackID, flashCard.Front, flashCard.Back });
+            }
+        }
+        catch
+        {
+            throw;
+        }
```
Comment (antonderegt) [inline Flashcards.DouglasFir/Flashcards/DAO/FlashCardDao.cs:38]: In your other code you handle exceptions really well with showing the error to the user for example. Here it's not doing anything. Behaviour will be the same without the try catch.

File: Flashcards.DouglasFir/Flashcards/DAO/FlashCardDao.cs [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+            FlashCard flashCard = new FlashCard
+            {
+                StackID = stack.StackID,
+                Front = flashCardDto.Front,
+                Back = flashCardDto.Back
+            };
+
+            using (var dbConnection = _dbContext.GetConnectionToFlashCards())
+            {
+                string sql = $"INSERT INTO {ConfigSettings.tbFlashCardsName} (StackID, Front, Back) VALUES (@StackID, @Front, @Back)";
```
Comment (antonderegt) [inline Flashcards.DouglasFir/Flashcards/DAO/FlashCardDao.cs:31]: I'm not really sure about how you set the table name. It's not user input, so low chance of sql injection, but I still think it'll be better to also inject the table name as a parameter.

File: Flashcards.DouglasFir/Flashcards/DAO/StudySessionDao.cs [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    public IEnumerable<dynamic>? GetAverageScoreReportData(int year)
+    {
+        try
+        {
+            using (var dbConnection = _dbContext.GetConnectionToFlashCards())
+            {
+                string sql = $@"
+                DECLARE @Year int = @SelectedYear;
+
+                ;WITH MonthlyAverages AS (
```
Comment (antonderegt) [inline Flashcards.DouglasFir/Flashcards/DAO/StudySessionDao.cs:106]: `;`

**Review by antonderegt — CHANGES_REQUESTED** [review]
Hey @rankdjr good job with all the requirements and extra challenges 🥳!

Unfortunately I'm going to ask you to change a few things. Firstly it doesn't run, I had to change the casing everywhere you used the `ConfigSettings` service. Secondly, when the database is empty the program crashes when I try to start a study session for example.

Please change those things and have a look at the comment I left in your code. Apart from those things you did a good job!

File: Flashcards.DouglasFir/Flashcards/App.config (hunk 1 above)
Comment (rankdjr) [reply Flashcards.DouglasFir/Flashcards/App.config:11]: Corrected on point 4. I was mixing conventions between sql and C#, but I do see the value in verbose names.

File: Flashcards.DouglasFir/Flashcards/App1.config (hunk 2 above)
Comment (rankdjr) [reply Flashcards.DouglasFir/Flashcards/App1.config:1]: Not sure how this snuck in there. This was removed early in development. Corrected and verified to be removed.

File: Flashcards.DouglasFir/Flashcards/App.config (hunk 1 above)
Comment (antonderegt) [reply Flashcards.DouglasFir/Flashcards/App.config:11]: @rankdjr I don't see a new comment yet, correct?

File: Flashcards.DouglasFir/Flashcards/DAO/FlashCardDao.cs (hunk 6 above)
Comment (rankdjr) [reply Flashcards.DouglasFir/Flashcards/DAO/FlashCardDao.cs]: I was changing the view/table names and was receiving an error so I went with the config file option. I reverted the flashcards file to use a hard coded table.

File: Flashcards.DouglasFir/Flashcards/App.config (hunk 1 above)
Comment (rankdjr) [reply Flashcards.DouglasFir/Flashcards/App.config]: @antonderegt correct. had a busy few weeks, so just getting around to finishing these changes

**Review by antonderegt — APPROVED** [review]

**Pull request closed without merging**